Release 23.7.0 (#12901)
conda/conda: CHANGELOG.md

* Mark `conda.models.leased_path_entry.LeasedPathEntry` as pending deprecation. (#12735)
* Mark `conda.models.enums.LeasedPathType` as pending deprecation. (#12735)
* Mark `conda.common.temporary_content_in_file` as pending deprecation. Use `tempfile` instead. (#12795)
* Mark `conda.cli.python_api` as pending deprecation. Use `conda.testing.conda_cli` fixture instead. (#12796)

### Docs

* Document how to use the new "pre_commands" plugin hook. (#12712)
* Document how to use the new "post_commands" plugin hook. (#12758)
* Add docstrings to all public modules. (#12792)
* Auto-generate API docs using `sphinx-autoapi`. (#12798)
* Convert all manual redirects into config using `sphinx-reredirects`. (#12798)
* Revise the plugins index page to make it easier to understand how to create a conda plugin. (#12802)
* Add missing `conda env` CLI docs. (#12841)

### Other

* Update `tests/cli/test_main_rename.py` to use latest fixtures. (#12517)
* Update `tests/test_activate.py` to test the new behavior. (#12769)
* Re-enable all `conda_env` tests and remove irrelevant tests. (#12813)
* Convert all `unittest`-style tests to `pytest`-style. (#12819)

> * Convert tests/test-recipes into local noarch packages instead of relying on conda-test channel and local builds. (#12879)

Author: ```suggestion
* Convert `tests/test-recipes` into local noarch packages instead of relying on conda-test channel and local builds. (#12879)
```

[//]: # (current developments)

> ## 23.7.0 (2023-07-17)

Reviewer: Please update to current date.

> ## 23.7.0 (2023-07-17)

Author: ```suggestion
## 23.7.0 (2023-07-25)
```

### Enhancements

* Add `conda.deprecations.DeprecationHandler.action` helper to deprecate `argparse.Action`s. (#12493)

> * Add support for FreeBSD. (#12647)

Reviewer: ```suggestion
* Add support for the FreeBSD operating system and register `freebsd-64` as a known subdirectory for FreeBSD on x86-64. (#12647)
```

* Do not mock `$CONDA_PREFIX` when `--name` or `--prefix` is provided. (#12696)
* Add support for `sha256` filters in the MatchSpec syntax (e.g. `*[sha256=f453db4ffe2271ec492a2913af4e61d4a6c118201f07de757df0eff769b65d2e]`). (#12654 via #12707)

> * Add a new "pre_commands" hook allowing plugins to run code before a `conda` subcommand. (#12712, #12864)
> * Add a new health check to `conda doctor` detecting altered packages in an environment by comparing expected and computed `sha256` checksums. (#12757)
> * Add a new "post_commands" hook allowing plugins to run code before a `conda` subcommand. (#12758, #12864)

Reviewer: ```suggestion
* Add a new health check to `conda doctor` detecting altered packages in an environment by comparing expected and computed `sha256` checksums. (#12757)
* Add new `pre_commands` and `post_commands` plugin hooks allowing plugins to run code before and after `conda` subcommands. (#12712, #12758, #12864)
```

> * Stop using `distutils` directly in favor of the vendored version in `setuptools` 60 and later. (#11136)

Reviewer: ```suggestion
* Stop using `distutils` directly in favor of the vendored version in `setuptools` 60 and later or standard library equivalents. (#11136)
```

* Add a `CITATION.cff` file to the root of the repository to make it easier for users to cite conda. (#12781)
* Add optional `CondaSubcommand.configure_parser` allowing third-party plugins to hook into conda's argument parser. (#12814)
* Only display third-party subcommands in `conda --help` and not for every other subcommand. (#12814, #12740)
* Add a new config option, `no_plugins`, a` --no-plugins` command line flag, and a `CONDA_NO_PLUGINS` environment variable that disables external plugins for built-in conda commands. (#12748)
* Register plugins using their canonical/fully-qualified name instead of the easily spoofable entry point name. (#12869)
* De-duplicate plugin and legacy subcommands in `conda --help`. (#12893)
* Implement a 2-phase parser to better handle plugin disabling (via `--no-plugins`). (#12910)

> * Refactor subcommand parsing to use a greedy parser since `argparse.REMAINDER` has known issues (https://github.com/python/cpython/issues/61252). (#12910)

Reviewer: ```suggestion
* Refactor subcommand parsing to use a greedy parser since `argparse.REMAINDER` has [known issues](https://github.com/python/cpython/issues/61252). (#12910)
```

* Don't export `__osx` virtual package when `CONDA_OVERRIDE_OSX=""`. (#12715)
* Fix erroneous `conda deactivate` behavior of unsetting preexisting environment variables that are identical to those set during `conda activate`. (#12769)
* Correct third-party subcommands to receive _remaining_ arguments instead of a blanket `sys.argv[2:]` which broke `conda_cli` testing. (#12814, #12910)

### Deprecations

* Mark `conda.base.context.context.root_dir` as pending deprecation. Use `conda.base.context.context.root_prefix` instead. (#12701)
* Mark `conda.plugins.subcommands.doctor.cli.get_prefix` as pending deprecation. Use `conda.base.context.context.target_prefix` instead. (#12725)
* Mark `conda.models.leased_path_entry.LeasedPathEntry` as pending deprecation. (#12735)
* Mark `conda.models.enums.LeasedPathType` as pending deprecation. (#12735)
* Mark `conda.common.temporary_content_in_file` as pending deprecation. Use `tempfile` instead. (#12795)
* Mark `conda.cli.python_api` as pending deprecation. Use `conda.testing.conda_cli` fixture instead. (#12796)

### Docs

> * Document how to use the new "pre_commands" plugin hook. (#12712)
> * Document how to use the new "post_commands" plugin hook. (#12758)

Reviewer: ```suggestion
* Document how to use the new `pre_commands` and `post_commands` plugin hook. (#12712, #12758)
```

[//]: # (current developments)

## 23.7.0 (2023-07-17)

### Enhancements

* Add `conda.deprecations.DeprecationHandler.action` helper to deprecate `argparse.Action`s. (#12493)
* Add support for FreeBSD. (#12647)
* Do not mock `$CONDA_PREFIX` when `--name` or `--prefix` is provided. (#12696)
* Add support for `sha256` filters in the MatchSpec syntax (e.g. `*[sha256=f453db4ffe2271ec492a2913af4e61d4a6c118201f07de757df0eff769b65d2e]`). (#12654 via #12707)
* Add a new "pre_commands" hook allowing plugins to run code before a `conda` subcommand. (#12712, #12864)
* Add a new health check to `conda doctor` detecting altered packages in an environment by comparing expected and computed `sha256` checksums. (#12757)
* Add a new "post_commands" hook allowing plugins to run code before a `conda` subcommand. (#12758, #12864)
* Stop using `distutils` directly in favor of the vendored version in `setuptools` 60 and later. (#11136)

> *   <news item>

Reviewer 2: ```suggestion
```

* Add a `CITATION.cff` file to the root of the repository to make it easier for users to cite conda. (#12781)
* Add optional `CondaSubcommand.configure_parser` allowing third-party plugins to hook into conda's argument parser. (#12814)
* Only display third-party subcommands in `conda --help` and not for every other subcommand. (#12814, #12740)
* Add a new config option, `no_plugins`, a` --no-plugins` command line flag, and a `CONDA_NO_PLUGINS` environment variable that disables external plugins for built-in conda commands. (#12748)
* Register plugins using their canonical/fully-qualified name instead of the easily spoofable entry point name. (#12869)
* De-duplicate plugin and legacy subcommands in `conda --help`. (#12893)

### Bug fixes

* Use `requests.exceptions.JSONDecodeError` for ensuring compatibility with different `json` implementations used by requests. This fixes a bug that caused only the first of multiple given source URLs to be tried. This also raises the minimum required requests version to 2.27.0. (#12683)
* Don't export `__osx` virtual package when `CONDA_OVERRIDE_OSX=""`. (#12715)

> *   Fix erroneous `conda deactivate` behavior of unsetting preexisting environment variables that are identical to those set during `conda activate`. (#12769)

Reviewer 2: ```suggestion
* Fix erroneous `conda deactivate` behavior of unsetting preexisting environment variables that are identical to those set during `conda activate`. (#12769)
```

* Correct third-party subcommands to receive _remaining_ arguments instead of a blanket `sys.argv[2:]` which broke `conda_cli` testing. (#12814)

### Deprecations

* Mark `conda.base.context.context.root_dir` as pending deprecation. Use `conda.base.context.context.root_prefix` instead. (#12701)
* Mark `conda.plugins.subcommands.doctor.cli.get_prefix` as pending deprecation. Use `conda.base.context.context.target_prefix` instead. (#12725)
* Mark `conda.models.leased_path_entry.LeasedPathEntry` as pending deprecation. (#12735)
* Mark `conda.models.enums.LeasedPathType` as pending deprecation. (#12735)

> *   <news item>

Reviewer 2: ```suggestion
```

* Mark `conda.common.temporary_content_in_file` as pending deprecation. Use `tempfile` instead. (#12795)
* Mark `conda.cli.python_api` as pending deprecation. Use `conda.testing.conda_cli` fixture instead. (#12796)

### Docs

* Document how to use the new "pre_commands" plugin hook. (#12712)
* Document how to use the new "post_commands" plugin hook. (#12758)

> *   <news item>

Reviewer 2: ```suggestion
```

* Add docstrings to all public modules. (#12792)
* Auto-generate API docs using `sphinx-autoapi`. (#12798)
* Convert all manual redirects into config using `sphinx-reredirects`. (#12798)
* Revise the plugins index page to make it easier to understand how to create a conda plugin. (#12802)
* Add missing `conda env` CLI docs. (#12841)

### Other

* Update `tests/cli/test_main_rename.py` to use latest fixtures. (#12517)

> *   Update `tests/test_activate.py` to test the new behavior. (#12769)

Reviewer 2: ```suggestion
* Update `tests/test_activate.py` to test the new behavior. (#12769)
```
